src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+
+    MouseSER_RegisterListener(this);
+}
+
+CSerialMouse::~CSerialMouse() {
+    MouseSER_UnRegisterListener(this);
+    removeEvent(SERIAL_TX_EVENT); // clear events
+    setType(MouseType::NO_MOUSE);
+}
+
+void CSerialMouse::setType(MouseType type) {
+    if (type != mouse_type) {
+        mouse_type = type;
+        const char* type_name = nullptr;
+        switch (type) {
+         case MouseType::NO_MOUSE: // just to print out log in the destructor
+            type_name            = "(none)";
+            break;
+        case MouseType::MICROSOFT:
+            type_name            = "Microsoft, 2 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = false;
+            mouse_has_wheel      = false;
+            break;
+        case MouseType::LOGITECH:
+            type_name            = "Logitech, 3 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = false;
+            break;
+        case MouseType::WHEEL:
+            type_name            = "wheel, 3 buttons";
+            mouse_bytelen        = 7;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = true;
+            break;
+        case MouseType::MOUSE_SYSTEMS:
+            type_name            = "Mouse Systems, 3 buttons";
+            mouse_bytelen        = 8;
+            mouse_has_3rd_button = true;
+            mouse_has_wheel      = false;
+            break;      
+        default:
+            unimplemented();
+            break;
+        }
+
+        if (type_name)
+            LOG_MSG("MOUSE (COM%d): %s", port_num, type_name);
+    }
+}
+
+void CSerialMouse::abortPacket() {
+    packet_len          = 0;
+    xmit_idx            = 0xff;
+    xmit_2part          = false;
+    xmit_another_move   = false;
+    xmit_another_button = false;
+}
+
+void CSerialMouse::clearCounters() {
+    mouse_delta_x       = 0;
+    mouse_delta_y       = 0;
+    mouse_delta_w       = 0;  
+}
+
+void CSerialMouse::mouseReset() {
+    abortPacket();
+    clearCounters();
+    send_ack = true;
+
+    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::onMouseEventMoved(int16_t delta_x, int16_t delta_y) {
+    mouse_delta_x += delta_x;
+    mouse_delta_y += delta_y;
+
+    // initiate data transfer and form the packet to transmit. if another packet
+    // is already transmitting now then wait for it to finish before transmitting ours,
+    // and let the mouse motion accumulate in the meantime
+
+    if (xmit_idx >= packet_len)
+        startPacketData();
+    else
+        xmit_another_move = true;
+}
+
+void CSerialMouse::onMouseEventButton(uint8_t buttons, uint8_t idx) {
+    mouse_buttons = buttons;
+
+    if (!mouse_has_3rd_button && idx > 1) return;
+
+    if (xmit_idx >= packet_len)
+        startPacketData(idx > 1);
+    else
+        xmit_another_button = true;
+}
+
+void CSerialMouse::onMouseEventWheel(int8_t delta_w) {
```

Comment:
General style comment for the whole class: the clang formatter puts opening braces on new lines for function/method declarations, so all methods should be updated accordingly.